### PR TITLE
bench: Expand mempool eviction benchmarking

### PR DESCRIPTION
**Pull request by instagibbs**
Took the existing benchmark and made is something more exhaustive, targeting eviction of 2500 transactions of different shapes. of dependencies.

Magic number comes from: 100 children being RBF'd, each bumping 24 parents = 2500 transactions

Turns out the benchmark `ComplexMemPool` already does something similar, so if we don't want to expand it, maybe just remove it in favor of the existing better benchmark?

### CODE REVIEW

File: src/bench/mempool_eviction.cpp [hunk 1]
```diff
@@ -21,116 +21,65 @@ static void AddTx(const CTransactionRef& tx, const CAmount& nFee, CTxMemPool& po
...
-    tx4.vin[1].scriptSig = CScript() << OP_4;
-    tx4.vin[1].scriptWitness.stack.push_back({4});
-    tx4.vout.resize(2);
-    tx4.vout[0].scriptPubKey = CScript() << OP_4 << OP_EQUAL;
-    tx4.vout[0].nValue = 10 * COIN;
-    tx4.vout[1].scriptPubKey = CScript() << OP_4 << OP_EQUAL;
-    tx4.vout[1].nValue = 10 * COIN;
+            if (put_index < txns.size()) {
+                // We spend put_index's txn's in the next available slot for each previous transaction
+                tx.vin[put_index].prevout = COutPoint(txns[put_index]->GetHash(), put_index + txn_index - 1);
```
Comment (MarcoFalke) [inline src/bench/mempool_eviction.cpp:42]:
```
bench/mempool_eviction.cpp:42:105: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'unsigned long'

**Comment by Collinschimaobi** [comment]
https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md

File: src/bench/mempool_eviction.cpp (hunk 1 above)
Comment (instagibbs) [reply src/bench/mempool_eviction.cpp:42]: you running this without any args? txn_index has to be greater than 0, so unsure how that would happen

Comment (instagibbs) [reply src/bench/mempool_eviction.cpp:43]: found the issue, thanks

File: src/bench/mempool_eviction.cpp [hunk 2]
```diff
@@ -21,116 +21,96 @@ static void AddTx(const CTransactionRef& tx, const CAmount& nFee, CTxMemPool& po
...
 }
 
-// Right now this is only testing eviction performance in an extremely small
-// mempool. Code needs to be written to generate a much wider variety of
-// unique transactions for a more meaningful performance measurement.
-static void MempoolEviction(benchmark::Bench& bench)
+void add_parents_child(std::vector<CTransactionRef>& txns, size_t package_size, size_t set_num)
+{
+    // N+1 N-input-N-output txns, in a parents-and-child package
+    // Where the final tx takes an input from all ancestors
```
Comment (glozow) [inline src/bench/mempool_eviction.cpp:27]: I could be reading this code wrong, but isn't each tx taking inputs from each tx before it? It's still parents-and-child technically but it's also a chain.

```suggestion
    // Where each tx takes an input from all prior txns
```

File: src/bench/mempool_eviction.cpp [hunk 3]
```diff
@@ -21,116 +21,96 @@ static void AddTx(const CTransactionRef& tx, const CAmount& nFee, CTxMemPool& po
...
+{
+    const auto testing_setup = MakeNoLogFileContext<const TestingSetup>();
 
+    std::vector<CTransactionRef> txns;
+
+    for (size_t i=0; i<num_packages; i++) {
+        add_parents_child(txns, package_size, i);
+    }
+
+    /* Now we're ready */
```
Comment (glozow) [inline src/bench/mempool_eviction.cpp:86]: Er, what does this comment mean?

File: src/bench/mempool_eviction.cpp [hunk 4]
```diff
@@ -21,116 +21,96 @@ static void AddTx(const CTransactionRef& tx, const CAmount& nFee, CTxMemPool& po
...
-    // Create transaction references outside the "hot loop"
-    const CTransactionRef tx1_r{MakeTransactionRef(tx1)};
-    const CTransactionRef tx2_r{MakeTransactionRef(tx2)};
-    const CTransactionRef tx3_r{MakeTransactionRef(tx3)};
-    const CTransactionRef tx4_r{MakeTransactionRef(tx4)};
-    const CTransactionRef tx5_r{MakeTransactionRef(tx5)};
-    const CTransactionRef tx6_r{MakeTransactionRef(tx6)};
-    const CTransactionRef tx7_r{MakeTransactionRef(tx7)};
+    bench.run([&]() NO_THREAD_SAFETY_ANALYSIS {
+        for (size_t i=0; i<txns.size(); i++) {
```
Comment (glozow) [inline src/bench/mempool_eviction.cpp:64]: nit: I think we tend to prefer this
```suggestion
        for (size_t i{0}; i < txns.size(); ++i) {
```

**Review by glozow — COMMENTED** [review]
Thanks for beefing up this bench! Basically lgtm, a few nits. Would like a bit clearer docs so future people can easily understand what the test is doing.

**Comment by instagibbs** [comment]
@jonatack good point, updated

File: src/bench/mempool_eviction.cpp [hunk 5]
```diff
@@ -21,116 +21,96 @@ static void AddTx(const CTransactionRef& tx, const CAmount& nFee, CTxMemPool& po
...
+            // Monotonically decreasing fees
+            AddTx(txns[i], (txns.size() - i)*1000LL, pool);
+        }
+        assert(pool.size() == txns.size());
+        pool.TrimToSize(pool.DynamicMemoryUsage() - 1);
+        assert(pool.size() == txns.size() - 1);
+        pool.TrimToSize(GetVirtualTransactionSize(*txns[0]));
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/bench/mempool_eviction.cpp:71]: hmmm, I don't recall. I can just set to 0

File: src/bench/mempool_eviction.cpp (hunk 3 above)
Comment (instagibbs) [reply src/bench/mempool_eviction.cpp:86]: will remove

File: src/bench/mempool_eviction.cpp [hunk 6]
```diff
@@ -21,116 +21,96 @@ static void AddTx(const CTransactionRef& tx, const CAmount& nFee, CTxMemPool& po
...
-        AddTx(tx2_r, 5000LL, pool);
-        AddTx(tx3_r, 20000LL, pool);
-        AddTx(tx4_r, 7000LL, pool);
-        AddTx(tx5_r, 1000LL, pool);
-        AddTx(tx6_r, 1100LL, pool);
-        AddTx(tx7_r, 9000LL, pool);
-        pool.TrimToSize(pool.DynamicMemoryUsage() * 3 / 4);
-        pool.TrimToSize(GetVirtualTransactionSize(*tx1_r));
+        for (size_t i=0; i<txns.size(); i++) {
+            // Monotonically decreasing fees
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/bench/mempool_eviction.cpp]: looks like something that doesn't matter, based on me getting rid of that factor. Removed.

**Comment by instagibbs** [comment]
fixups included, squashed

File: src/bench/mempool_eviction.cpp (hunk 3 above)
Comment (glozow) [reply src/bench/mempool_eviction.cpp:85]: Didn't remove?

File: src/bench/mempool_eviction.cpp [hunk 7]
```diff
@@ -21,116 +21,94 @@ static void AddTx(const CTransactionRef& tx, const CAmount& nFee, CTxMemPool& po
...
+{
+    // N+1 N-input-N-output txns, in a parents-and-child package
+    // Where each tx takes an input from all prior txns
+    const size_t num_puts = std::max(package_size - 1, (size_t)1);
+    const size_t num_txns = package_size;
+    for (size_t txn_index=0; txn_index<num_txns; txn_index++) {
+        CMutableTransaction tx = CMutableTransaction();
+        tx.vin.resize(num_puts);
+        tx.vout.resize(num_puts);
+        for (size_t put_index=0; put_index<num_puts; put_index++) {
```
Comment (glozow) [inline src/bench/mempool_eviction.cpp:34]: Can you apply the changes to the other loops too? i.e.
```suggestion
        for (size_t put_index{0}; put_index < num_puts; ++put_index) {
```

File: src/bench/mempool_eviction.cpp [hunk 8]
```diff
@@ -21,116 +21,94 @@ static void AddTx(const CTransactionRef& tx, const CAmount& nFee, CTxMemPool& po
...
+            tx.vin[put_index].scriptSig = CScript() << CScriptNum(txn_index);
+            tx.vin[put_index].scriptWitness.stack.push_back({1});
+            tx.vout[put_index].scriptPubKey = CScript() << CScriptNum(set_num); // Ensures unique txid with set_num
+            tx.vout[put_index].nValue = txn_index;
+
+            if (put_index < txn_index) {
+                // We spend put_index's txn's in the next available slot for each previous transaction
+                assert(put_index + txn_index >= 1);
+                tx.vin[put_index].prevout = COutPoint(txns[put_index]->GetHash(), put_index + txn_index - 1);
```
Comment (glozow) [inline src/bench/mempool_eviction.cpp:43]: Wait, if you're using `txns[put_index]` *and* you're using the same `txns` reference each time you call `add_parents_child`, aren't you having all subsequent packages spend outputs from txns[0:24] too? That means all these packages conflict with each other.

You could fix this by adding `set_num * package_size` or something to the index. But what would probably make a better interface is if `add_parents_child` returns the list of transactions it creates, and you append them to your larger list of packages.

Comment (instagibbs) [reply src/bench/mempool_eviction.cpp:43]: Hmmm, traced out a package size of 4, looks like I'm referencing non-existent outputs, not re-using them(maybe both :) ).

Clearly this is confusing and broken either way.

To take a step back, after doing this I realized https://github.com/bitcoin/bitcoin/blob/master/src/bench/mempool_stress.cpp exists. I'm fine if this doesn't get merged if it doesn't add anything new? Helped me learn about the benchmarks if nothing else. Thoughts?

Comment (instagibbs) [reply src/bench/mempool_eviction.cpp:43]: I could change the PR to remove this silly bench.. :P

**Pull request closed without merging**